blockindex/indexer_group.go

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) 2023 IoTeX Foundation
+// This source code is provided 'as is' and no warranties are given as to title or non-infringement, merchantability
+// or fitness for purpose and, to the extent permitted by law, all liability for your use of the code is disclaimed.
+// This source code is governed by Apache License 2.0 that can be found in the LICENSE file.
+
+package blockindex
+
+import (
+	"context"
+
+	"github.com/iotexproject/iotex-core/blockchain/block"
+	"github.com/iotexproject/iotex-core/blockchain/blockdao"
+)
+
+// IndexerGroup is a special index that includes multiple indexes,
+// which stay in sync when blocks are added.
+type IndexerGroup struct {
+	indexers []blockdao.BlockIndexer
+}
+
+// NewIndexerGroup creates a new indexer group
+func NewIndexerGroup(indexers ...blockdao.BlockIndexer) *IndexerGroup {
+	return &IndexerGroup{indexers: indexers}
+}
+
+// Start starts the indexer group
+func (ig *IndexerGroup) Start(ctx context.Context) error {
+	for _, indexer := range ig.indexers {
+		if err := indexer.Start(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Stop stops the indexer group
+func (ig *IndexerGroup) Stop(ctx context.Context) error {
+	for _, indexer := range ig.indexers {
+		if err := indexer.Stop(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// PutBlock puts a block into the indexers in the group
+func (ig *IndexerGroup) PutBlock(ctx context.Context, blk *block.Block) error {
+	for _, indexer := range ig.indexers {
+		// check if the indexer is a BlockIndexerWithStart
+		if indexerWithStart, ok := indexer.(blockdao.BlockIndexerWithStart); ok {
```

Comment:
use `startHeight[i]`, so no need to do the type insertion every time

Reply:
Here is benchmark for type assert:
```go
func BenchmarkTypeAssert(b *testing.B) {
	ctrl := gomock.NewController(b)
	var indexer interface{}
	indexer = mock_blockdao.NewMockBlockIndexer(ctrl)
	b.ResetTimer()

	b.Run("type assert", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			_, _ = indexer.(blockdao.BlockIndexerWithStart)
		}
	})
}
```

And the result:
```
❯ go test -benchmem -run=^$ -bench ^BenchmarkTypeAssert$ github.com/iotexproject/iotex-core/blockindex
goos: darwin
goarch: arm64
pkg: github.com/iotexproject/iotex-core/blockindex
BenchmarkTypeAssert/type_assert-10              344110524                3.417 ns/op           0 B/op          0 allocs/op
PASS
ok      github.com/iotexproject/iotex-core/blockindex   1.870s
```

I think the `3.4ns` is acceptable for function calls, so there is no need to introduce more complex logic to optimize at this point

Reply:
1. the benchmark does not have `if`, also `_, _` means the result is not used, go compiler may optimize to not call the right side at all (but I am not sure about that for bench-marking code)
2. it's not all about performance. When a group is created, the start heights are known at creation time. If something can be done just one-time at creation, why keep doing it in a `for` loop for thousands of times ... To me the current way is complex, b/c there is a much simpler/clean way of just doing it one-time